sqlglot/parser.py

```diff
@@ -1585,8 +1586,13 @@ def _parse(
         total = len(raw_tokens)
         chunks: t.List[t.List[Token]] = [[]]
 
+        block_nesting_level = 0
         for i, token in enumerate(raw_tokens):
-            if token.token_type == TokenType.SEMICOLON:
+            if self._is_nested_block_start(token, i, raw_tokens):
```

Comment:
Note: Curious to get your thoughts on this. This is an attempt to determine whether we are executing in the context of a nested code block (i.e., `BEGIN ... END`) and to preserve semicolon tokens within the current chunk. 

The idea of `_is_nested_block_start(..)` is to let dialects decide whether we are executing within a code block where semicolons should be preserved. By default the method returns `False`, and the `snowflake` dialect parser overwrites this method.

There may of course be better / more elegant ways to achieve that :) - happy for any suggestions!